### PR TITLE
bpo-21910: Clarified codecs writelines method

### PR DESCRIPTION
**Pull request by slateny**
See [issue](https://bugs.python.org/issue21910). [`writelines`](https://github.com/python/cpython/blob/main/Lib/codecs.py#L385) uses `.join()` instead of some streaming implementation so for infinite / very large generators the method will freeze. Updated documentation so it reflects this behavior

https://bugs.python.org/issue21910

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@slateny

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Doc/library/codecs.rst [hunk 1]
```diff
@@ -697,8 +697,8 @@ compatible with the Python codec registry.
 
    .. method:: writelines(list)
 
-      Writes the concatenated list of strings to the stream (possibly by reusing
-      the :meth:`write` method). The standard bytes-to-bytes codecs
+      Writes the concatenated list of strings to the stream by reusing
+      the :meth:`write` method, and does not support infinite or very large generators. The standard bytes-to-bytes codecs
```
Comment (JelleZijlstra) [inline Doc/library/codecs.rst]: This doesn't read well. Should put it in a separate sentence instead.

Comment (slateny) [reply Doc/library/codecs.rst]: How does this sound instead?

```
Writes the concatenated list of strings to the stream. Reuses the write method, so infinite or very large generators are not supported. The standard . . .
```

Comment (JelleZijlstra) [reply Doc/library/codecs.rst]: That's not quite right; the reason long generators aren't supported isn't that it uses the write methods, but that it writes everything at once.

I'd reword the first line to say "iterable" instead of "list" and say that "Infinite or very long iterables are not supported."

Comment (slateny) [reply Doc/library/codecs.rst]: Ah, that's a good point, then would this work?

```
Concatenates the string iterable and writes it to the stream. Uses .join() for concatenation, so infinite or very long iterables are not supported.
```

Comment (JelleZijlstra) [reply Doc/library/codecs.rst]: That is correct, but it seems like the docstring goes out of its way to avoid saying too much about how the method is implemented, so maybe we should keep it less concrete.

Comment (slateny) [reply Doc/library/codecs.rst]: But the comment on not supporting large iterables is based off using `.join`, so wouldn't this be one of the cases where it'd be nice to go very briefly into implementation?

Comment (akulakov) [reply Doc/library/codecs.rst]: I agree with Jelle about leaving out implementation detail because the `join()` wouldn't be an issue if it was buffering the lines, - so this can be confusing to readers. It can also be confusing to some readers because `join()` is recommended as the proper operation for concatenation as opposed to `+=`. You would have to go down the rabbit hole of explaining more and more details about implementation (which can also change in the future).

Comment (slateny) [reply Doc/library/codecs.rst]: Hmm interesting, so we would leave out the implementation, and if readers are curious on why infinite/large generators are not supported they would just read the source code themselves?

Comment (JelleZijlstra) [reply Doc/library/codecs.rst]: Correct. The docstring here basically provides a guarantee to readers, so if we are too precise, that constrains how we can change the function in the future.

File: Doc/library/codecs.rst [hunk 2]
```diff
@@ -697,8 +697,8 @@ compatible with the Python codec registry.
 
    .. method:: writelines(list)
 
-      Writes the concatenated list of strings to the stream (possibly by reusing
-      the :meth:`write` method). The standard bytes-to-bytes codecs
+      Concatenates the string iterable and writes it to the stream. Infinite or
```
Comment (JelleZijlstra) [inline Doc/library/codecs.rst]: I don't like the way the rest of the docs was reworded. It's not clear to a reader what "the string iterable" is. Let's just leave the first sentence unchanged.

Comment (slateny) [reply Doc/library/codecs.rst]: Is `Writes the concatenated iterable of strings to the stream` better or `Writes the concatenated string iterable to the stream`?

Comment (JelleZijlstra) [reply Doc/library/codecs.rst]: I prefer the first one; it reads more clearly.

**Comment by slateny** [comment]
@JelleZijlstra Could you take another peek and let me know if it looks good to you?

File: Doc/library/codecs.rst [hunk 3]
```diff
@@ -697,8 +697,8 @@ compatible with the Python codec registry.
 
    .. method:: writelines(list)
 
-      Writes the concatenated list of strings to the stream (possibly by reusing
-      the :meth:`write` method). The standard bytes-to-bytes codecs
+      Writes the concatenated iterable of strings to the stream. Infinite or
+      very large iterables are not supported. The standard bytes-to-bytes codecs
```
Comment (JelleZijlstra) [inline Doc/library/codecs.rst]:
```suggestion
      Writes the concatenated iterable of strings to the stream (possibly by reusing
      the :meth:`write` method). Infinite or
      very large iterables are not supported. The standard bytes-to-bytes codecs
```
No reason to remove this

Comment (slateny) [reply Doc/library/codecs.rst]: But wasn't it the case that we want to avoid specifying implementation details? And, while accurate, I'm not too sure what vague wording like 'possibly' brings, as (to me) it doesn't seem to help much

Comment (JelleZijlstra) [reply Doc/library/codecs.rst]: We also want to minimize changes. I'd rather keep this change focused on the minimum needed to fix the original bug.

Comment (slateny) [reply Doc/library/codecs.rst]: Sure, that sounds good to me. Do you think the parenthesized phasing should be removed though? If so, I could open up a separate PR with just that change, otherwise I'll just leave it be.

Comment (JelleZijlstra) [reply Doc/library/codecs.rst]: I don't think it's worth changing.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
@gvanrossum planning to merge this docs change.
